apps/dash-stitching/app.py

```diff
@@ -0,0 +1,483 @@
+        ),
+    ],
+    className="row twelve columns",
+)
+
+
+@app.callback(
+    Output("tabs-content-example", "children"), [Input("stitching-tabs", "value")]
+)
+def fill_tab(tab):
+    if tab == "canvas-tab":
+        return [
+            dash_canvas.DashCanvas(
+                id="canvas-stitch",
+                width=canvas_width,
+                height=canvas_height,
+                scale=scale,
+                lineWidth=2,
+                lineColor="red",
+                tool="line",
+                hide_buttons=["pencil"],
+                image_content=array_to_data_url(
+                    np.zeros((height, width), dtype=np.uint8)
+                ),
+                goButtonTitle="Estimate translation",
+            ),
+            html.Div(
+                children=[html.Button("Upload demo data")],
+                id="demo",
+                className="demo_button",
+            ),
+        ]
+    elif tab == "result-tab":
+        return [
+            dcc.Loading(
+                id="loading-1",
+                children=[
+                    html.Img(
+                        id="stitching-result",
+                        src=array_to_data_url(
+                            np.zeros((height, width), dtype=np.uint8)
+                        ),
+                        width=canvas_width,
+                    )
+                ],
+                type="circle",
+            ),
+            html.Div(
+                [
+                    html.Label("Contrast"),
+                    dcc.Slider(
+                        id="contrast-stitch", min=0, max=1, step=0.02, value=0.5
+                    ),
+                ],
+                # style={'width': '70%'},
+                className="result_slider",
+            ),
+            html.Div(
+                [
+                    html.Label("Brightness"),
+                    dcc.Slider(
+                        id="brightness-stitch", min=0, max=1, step=0.02, value=0.5
+                    ),
+                ],
+                # style={'width': '70%'},
+                className="result_slider",
+            ),
+        ]
+    else:
+        return [
+            html.Img(
+                id="bla", src=app.get_asset_url("stitch_demo.gif"), width=canvas_width
+            )
+        ]
+
+
+@app.callback(Output("stitching-tabs", "value"), [Input("button-stitch", "n_clicks")])
+def change_focus(click):
+    if click:
+        return "result-tab"
+    return "canvas-tab"
+
+
+@app.callback(Output("table-stitch", "data"), [Input("canvas-stitch", "json_data")])
+def estimate_translation(string):
+    props = parse_jsonstring_line(string)
+    if props is not None and len(props) > 0:
+        df = pd.DataFrame(props, columns=list_columns)
+        return df.to_dict("records")
+    else:
+        raise PreventUpdate
+
+
+@app.callback(Output("memory-stitch", "data"), [Input("button-stitch", "n_clicks")])
+def update_store(click):
+    sleep(0.1)
```

Comment:
Why do you have a `sleep` function here? Also you store something into `memory-stitch` but then never look at the stored data again.

Reply:
I'm not sure why the original author implemented the `sleep` function...

Reply:
@ronniebugia Might have something to do with some sort of race condition. @christopherjeon Could you try removing it and seeing if it still works?

Reply:
I think it still works